cpp/src/arrow/array_all.h

```diff
@@ -0,0 +1,28 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "arrow/array.h"                  // IWYU pragma: export
```

Comment:
It seems like `array_all.h` should be `array.h` and what is here array.h should be array/array_base.h or something

Reply:
If so, then it means there's not much benefit in splitting header files (except possibly readability). Everyone will intuitively include `array.h` and `builder.h`.

Reply:
Would maybe not including this these files be too onerous for developers?

Another advantage of more fine-grained header files is the potentially the elimination of "not real cycles" between compilation units I mentioned on the bazel thread.

Reply:
My experience in Python would be that finer-grained includes / imports increase the likelihood of cycles. For example, `array_list.h` and `array_binary.h` might have a cyclic dependency between each other while a monolithic `array.h` would not have a cycle onto itself (of course, this is a fictional example).